modules/webgl/src/context/context.js

```diff
@@ -200,27 +203,38 @@ function getVersion(gl) {
 // use devicePixelRatio to set canvas width and height
 function setDevicePixelRatio(gl, devicePixelRatio, options) {
   let devicePixelRatioClamped = false;
-  let aspectRatioValid = false;
+  let logWarning = false;
 
   // NOTE: if options.width and options.height not used remove in v8
   const clientWidth =
     'width' in options ? options.width : gl.canvas.clientWidth || gl.canvas.width || 1;
   const clientHeight =
     'height' in options ? options.height : gl.canvas.clientHeight || gl.canvas.height || 1;
 
-  // Note: when devicePixelRatio is too high, it is possible we might hit system limit for
-  // drawing buffer width and hight, in those cases they get clamped and resulting aspect ration may not be maintained
-  // for those cases, reduce devicePixelRatio.
-  do {
-    gl.canvas.width = Math.ceil(clientWidth * devicePixelRatio);
-    gl.canvas.height = Math.ceil(clientHeight * devicePixelRatio);
-    aspectRatioValid =
-      gl.drawingBufferWidth / clientWidth === gl.drawingBufferHeight / clientHeight;
-    devicePixelRatio = Math.max(devicePixelRatio / 2, 1);
-    devicePixelRatioClamped = devicePixelRatioClamped || !aspectRatioValid;
-  } while (!aspectRatioValid);
-
-  if (devicePixelRatioClamped) {
-    log.warn(`Device pixel ratio clamped`)();
+  const cachedSize = gl._canvasSizeInfo;
+  // Check if canvas needs to be resized
+  if (
+    cachedSize.clientWidth !== clientWidth ||
+    cachedSize.clientHeight !== clientHeight ||
+    cachedSize.devicePixelRatio !== devicePixelRatio
+  ) {
+    // Note: when devicePixelRatio is too high, it is possible we might hit system limit for
+    // drawing buffer width and hight, in those cases they get clamped and resulting aspect ration may not be maintained
+    // for those cases, reduce devicePixelRatio.
+    let clampedPixelRatio = devicePixelRatio;
+    do {
+      const canvasWidth = Math.floor(clientWidth * clampedPixelRatio);
+      const canvasHeight = Math.floor(clientHeight * clampedPixelRatio);
+      gl.canvas.width = canvasWidth;
+      gl.canvas.height = canvasHeight;
+      devicePixelRatioClamped =
+        gl.drawingBufferWidth !== canvasWidth || gl.drawingBufferHeight !== canvasHeight;
+      clampedPixelRatio = Math.max(clampedPixelRatio / 2, 1);
```

Comment:
I do not like this `while` loop:

- It's difficult to gauge its perf impact
- Do you know for sure that `clampedPixelRatio: 1` will definitely exit the loop?

If you set `clampedPixelRatio` to `Math.min(gl.drawingBufferWidth / clientWidth, gl.drawingBufferHeight / clientHeight)` you should not have to check clamping again.

Reply:
> Do you know for sure that `clampedPixelRatio: 1` will definitely exit the loop?

Based on the testing so for yes it does, but I couldn't find any documentation about how drawing buffer width/height are calculated (probably implementation dependent).

I can update the code to end the loop when `clampedPixelRatio` is `1` in addition to existing check.

> If you set clampedPixelRatio to Math.min(gl.drawingBufferWidth / clientWidth, gl.drawingBufferHeight / clientHeight) you should not have to check clamping again.

I am assuming, you mean, use `Math.min(gl.drawingBufferWidth / clientWidth, ...)` once and not perform the loop. The problem is aspect ratio, when width and height are clamped, aspect ratio is not maintain, hence single pixel ratio (drawingBufferWidth/clientWidth or  drawingBufferHeight/clientHeight) is not enough to map from CSS pixel to Device pixel.

Reply:
After you set pixel ratio to `Math.min(gl.drawingBufferWidth / clientWidth, ...)` you still need to set your canvas width and height using the new pixel ratio. They should not be clamped again because neither width and height would be larger than the last clamped values.